File: knapsack_benchmark/knapsack.cpp

#if __cplusplus < 201400L
#   error "need c++14 or later"
#endif

#define NDEBUG

#include <algorithm>
#include <chrono>
#include <cstdint>
#include <iostream>
#include <memory>
#include <tuple>
#include <vector>

//----------------------------------------------------------------------------
/*
    this is an almost one-for-one translation of the Julia version, see knapsack.jl
    for more comments.

    running this benchmark from command line:

    >g++ -O3 -Wall -o knapsack knapsack.cpp
    >./knapsack
*/
namespace knapsack
{
using std::int64_t;
using std::uint64_t;

//............................................................................

struct item
{
    int64_t const value;
    int64_t const weight;

}; // end of class
//............................................................................

int64_t
opt_value (int64_t const W, std::vector<item> const & items)
{
    auto const n = items.size ();

    std::unique_ptr<int64_t []> const v_data { std::make_unique<int64_t []> (W) }; // "zeros"
    std::unique_ptr<int64_t []> const v_prev_data { std::make_unique<int64_t []> (W) }; // "zeros"

    int64_t * V { v_data.get () };
    int64_t * V_prev { v_prev_data.get () };

    for (int64_t w = items[0].weight; w <= W; ++ w)
        V[w - 1] = items[0].value;

    for (std::size_t j = 1; j < n; ++ j)
    {
        std::swap (V, V_prev);
        item const & item { items [j] };

        for (int64_t w = 1; w <= W; ++ w)
        {
            auto const V_without_item_j = V_prev[w - 1];
            auto const V_allow_item_j = (w < item.weight
                ? V_without_item_j
                : (item.value + (w != item.weight ? V_prev[w - 1 - item.weight] : 0)));

            V[w - 1] = std::max(V_allow_item_j, V_without_item_j);
         }
    }

    return V[W - 1];
}
//............................................................................

uint64_t
next_rand (uint64_t & x) // note: 'x' maintains RNG state
{
    x ^= (x << 13);
    x ^= (x >> 7);
    x ^= (x << 17);

    return x;
}

using problem   = std::tuple<int64_t, std::vector<item>>;

problem
make_random_data (int64_t const W, int64_t const seed)
{
    int64_t const n = W / 100;
    uint64_t rng = seed;

    std::vector<item> items { };
    for (int64_t i = 0; i < n; ++ i)
    {
        int64_t const v = next_rand(rng) % 1000;
        int64_t const w = 1 + next_rand(rng) % (2 * W);

        items.emplace_back (item {v, w});
    }

    return std::make_tuple (W, std::move(items));
}
//............................................................................

void
run (int32_t const repeats = 5)
{
    std::vector<double> times(repeats);
    int64_t seed = 12345;

    for (int64_t W : { 5'000, 10'000, 20'000, 40'000, 80'000 })
    {
        for (int32_t repeat = 0; repeat < repeats; ++ repeat)
        {
            auto spec = make_random_data(W, seed += 1);
            auto start = std::chrono::high_resolution_clock::now();

            opt_value(std::get<0>(spec), std::get<1>(spec));

            auto stop = std::chrono::high_resolution_clock::now();

            times[repeat] = std::chrono::duration<double>(stop - start).count();
//            std::cout << "V = " << V << ", time = " << times[repeat] << std::endl;
        }

        std::sort (times.begin(), times.end());
        std::cout << "c++, " << W << ", " << times[repeats / 2] << std::endl;
    }
}

} // end of 'knapsack'
//----------------------------------------------------------------------------

int main()
{
    knapsack::run ();

	return 0;
}
//----------------------------------------------------------------------------

File: knapsack_benchmark/Knapsack.jl
__precompile__()
# ----------------------------------------------------------------------------
module Knapsack
#=
    see also:
        - the walkthrough at http://vladium.com/tutorials/study_julia_with_me/knapsack_benchmark/
        - https://en.wikipedia.org/wiki/Knapsack_problem#0/1_knapsack_problem

    running this benchmark from command line:

    >julia knapsack.jl
=#
# ............................................................................
"""
    we need a couple of integers to describe a knapsack item, its value and weight.
    a tuple would work, but a struct makes for more readable code
"""
struct Item
    value   ::Int64; # actually, being so type-specific is not necessary in Julia
    weight  ::Int64;
end
# ............................................................................
"""
    first version of the knapsack solver; uses a full matrix for V and will
    not scale to large problems
"""
function opt_value_BAD(W ::Int64, items ::Array{Item}) ::Int64
    n = length(items)

    # V[w,j] stores opt value achievable with capacity 'w' and using items '1..j':

    V = Array{Int64}(undef, W, n) # Wxn matrix with uninitialized storage

    # initialize first column v[:, 1] to trivial single-item solutions:
    # (note "broadcast assignment" syntax)

    V[:, 1] .= 0
    V[items[1].weight:end, 1] .= items[1].value

    # do a pass through remaining columns:

    for j in 2 : n
        itemⱼ = items[j]
        for w in 1 : W
            V_without_itemⱼ = V[w, j - 1]
            V_allow_itemⱼ = (w < itemⱼ.weight
                ? V_without_itemⱼ
                : (itemⱼ.value + (w ≠ itemⱼ.weight ? V[w - itemⱼ.weight, j - 1] : 0)))
            V[w, j] = max(V_allow_itemⱼ, V_without_itemⱼ)
        end
    end

    return V[W, n]
end
# ............................................................................
"""
    second version of the knapsack solver; uses two column buffers for V
"""
function opt_value(W ::Int64, items ::Array{Item}) ::Int64
    n = length(items)

    # V[w] stores opt value achievable with capacity 'w' and using items '1..j':

    V = zeros(Int64, W) # single column of size W, zero-initialized
    V_prev = Array{Int64}(undef, W) # single column of size W, uninitialized storage

    # initialize 'v' to trivial single-item solutions:
    # (note "broadcast assignment" syntax)

    V[items[1].weight:end] .= items[1].value

    # do a pass through remaining columns:

    for j in 2 : n
        V, V_prev = V_prev, V
        itemⱼ = items[j]
        for w in 1 : W
            V_without_itemⱼ = V_prev[w]
            V_allow_itemⱼ = (w < itemⱼ.weight
                ? V_without_itemⱼ
                : (itemⱼ.value + (w ≠ itemⱼ.weight ? V_prev[w - itemⱼ.weight] : 0)))
            V[w] = max(V_allow_itemⱼ, V_without_itemⱼ)
        end
    end

    return V[W]
end
# ............................................................................
"""
    a simple but good quality [xorshift RNG](https://en.wikipedia.org/wiki/Xorshift)
    that can be ported to all benchmark languages to ensure equivalent test problem
    instances across all ports
"""
function xorshift_rand(seed)
    @assert seed ≠ 0
    x ::UInt64 = seed # want unsigned bit arithmetic

    return function _next()
        x ⊻= (x << 13) # note that python would need 'nonlocal' here
        x ⊻= (x >>> 7)
        x ⊻= (x << 17)
        return x
    end
end

Problem = Tuple{Int64, Array{Item}} # use a type alias for less typing 

"""
    generate a randomized knapsack problem definition of scale 'W'
"""
function make_random_data(W ::Int64, seed ::Int64) ::Problem
    @assert W > 1000 # don't test on very small problems
    n = W ÷ 100
    rng = xorshift_rand(seed)

    items = Array{Item}(undef, n)
    for i in 1 : n
        v = rng() % 1000
        w = 1 + rng() % 2W
        items[i] = Item(v, w)
    end

    return W, items
end
# ............................................................................
"""
    you can use this function to load any knapsack problem defition from a file
    instead of using make_random_data();
    the format is:

     1. first line: <W>
     2. n subsequent lines: <value> <weight> (separated by spaces)
"""
function load_data(in_file ::String) ::Problem
    W = -1
    items = Item[]
    open(in_file) do fd
        for line in eachline(fd)
            line = strip(line)
            if isempty(line)
                break
            end
            if W < 0 # reading first line
                W = parse(Int, line)
            else # reading items
                tokens = split(line)
                @assert length(tokens) == 2
                item = Item(parse(Int, tokens[1]), parse(Int, tokens[2]))

                push!(items, item)
            end
        end
    end

    return W, items
end
# ............................................................................

function run(repeats = 5)
    @assert repeats > 1

    times = zeros(Float64, repeats)
    seed = 12345

    for W in [5_000, 10_000, 20_000, 40_000, 80_000]
        for repeat in 1 : repeats
            W, items = make_random_data(W, seed += 1)
            times[repeat] = @elapsed V = opt_value(W, items)
            # println("V = $(V), time = $(times[repeat])")
        end
        sort!(times)
        println("julia, ", W, ", ", times[(repeats + 1) ÷ 2]) # report W and median time
    end
end

# comment this out if you don't want to run the benchmark on module load:

run()

end # of module
# ----------------------------------------------------------------------------
